tutorial/lesson_21_auto_scheduler_generate.cpp

```diff
@@ -0,0 +1,192 @@
+                .estimate(z, 0, 64)
+                .estimate(n, 0, 4);
+
+            output2.estimate(x, 0, 128)
+                .estimate(y, 0, 128)
+                .estimate(z, 0, 64)
+                .estimate(n, 0, 4);
+
+            // Technically, the estimate values can be anything, but the closer
+            // they are to the actual use-case values, the better the generated
+            // schedule will be.
+
+            // Now, let's auto-schedule the pipeline by calling auto_schedule_outputs(),
+            // which takes in a MachineParams object as an argument. The machine_params
+            // argument is optional. If none is specified, the default machine parameters
+            // for a generic CPU architecture are going to be used by the auto-scheduler.
+
+            // Let's use some arbitrary but plausible values for the machine parameters.
+            const int kParallelism = 32;
+            const int kLastLevelCacheSize = 16 * 1024 * 1024;
+            const int kBalance = 40;
+            MachineParams machine_params(kParallelism, kLastLevelCacheSize, kBalance);
+            // The arguments to MachineParams are the maximum level of parallelism
+            // available, the size of the last-level cache (in KB), and the ratio
+            // between the cost of a miss at the last level cache and the cost
+            // of arithmetic, of the target architecture, in that order.
+
+            // Note that when using the auto-scheduler, no schedule should have
+            // been applied to the pipeline; otherwise, the auto-scheduler will
+            // throw an error. The current auto-scheduler does not work with
+            // partially-scheduled pipeline.
+            //
+            // Calling auto_schedule_outputs() will apply the generated schedule
+            // automatically to members of the pipeline in addition to returning
+            // a string representation of the schedule.
+            std::string schedule = auto_schedule_outputs(machine_params);
+            std::cout << "\nSchedule:\n\n" << schedule << "\n";
+
+            // The generated schedule that is dumped to std::cout is an actual
+            // Halide C++ source, which is readily copy-pastable back into
+            // this very same source file with little modifications. Programmers
+            // can use this as a starting schedule and iteratively improve the
+            // schedule. Note that the current auto-scheduler is only able to
+            // generate CPU schedule and only does tiling, simple vectorization
+            // and parallelization. It doesn't deal with line buffering, storage
+            // reordering, or factoring a reduction.
+        } else {
+            // This is where you would declare the schedule you have written by
+            // hand or paste the schedule generated by the auto-scheduler.
+
+            // This auto-scheduler will return the following schedule for the
```

Comment:
This is probably going to change, so prefix with:

// At the time of writing, the auto-scheduler will return the following ...

Reply:
Actually I would move this entire output comment block up into the if auto_schedule section, and just put here:

  // This is where you would declare the schedule you have written by
  // hand or paste the schedule generated by the auto-scheduler. We will use a naive schedule here to compare the performance of the autoschedule with a basic schedule.

and then just make everything compute_root(). Inlining is silly for this pipeline.